Add Hook to store arbitrary python objects that are copied over in tls

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92851
* __->__ #89169

For the cudagraphs implementation, we would like to reuse objects that are defined in python across the forward and backward. The backward is run in a different thread, so to handle this we add an api for copying over arbitrary python objects in pytorch's thread local state, in the same way that C++ objects are copied over currently.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

eellison committed:
Add Hook to store arbitrary python objects that are copied over in tls

[ghstack-poisoned]

eellison committed:
Update on "Add Hook to store arbitrary python objects that are copied over in tls"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Hook to store arbitrary python objects that are copied over in tls"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Hook to store arbitrary python objects that are copied over in tls"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -1500,6 +1501,23 @@ Call this whenever a new thread is created in order to propagate values from
+    return py::handle(obj);
+  });
+
+  py_module.def("_key_in_tls", [](std::string key) -> bool {
```
nit: `_is_key_in_tls`

albanD commented on aten/src/ATen/ThreadLocalPythonObjects.h:
```diff
@@ -0,0 +1,24 @@
+  static void set(std::string key, std::shared_ptr<SafePyObject> value);
+  static const std::shared_ptr<SafePyObject>& get(std::string key);
+  static bool contains(std::string key);
+  static int64_t size();
```
This is not actually used, you can remove it?

albanD reviewed (approved):
Implementation sounds good even though I don't really know if this is the right solution to your problem as it is not described in the PR description.

ezyang commented:
We are still doing this right?

eellison committed:
Update on "Add Hook to store arbitrary python objects that are copied over in tls"




cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Hook to store arbitrary python objects that are copied over in tls"


For the cudagraphs implementation, we would like to reuse objects that are defined in python across the forward and backward. The backward is run in a different thread, so to handle this we add an api for copying over arbitrary python objects in pytorch's thread local state, in the same way that C++ objects are copied over currently.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Add Hook to store arbitrary python objects that are copied over in tls"


For the cudagraphs implementation, we would like to reuse objects that are defined in python across the forward and backward. The backward is run in a different thread, so to handle this we add an api for copying over arbitrary python objects in pytorch's thread local state, in the same way that C++ objects are copied over currently.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
